includes/razorpay-webhook.php

```diff
@@ -110,4 +127,114 @@ protected function paymentAuthorized($data)
 
         exit;
     }
+
+    /**
+     * Currently we handle only subscription failures using this webhook
+     *
+     * @param $data
+     */
+    protected function paymentFailed($data)
+    {
+        //
+        // Order entity should be sent as part of the webhook payload
+        //
+        $orderId = $data['payload']['payment']['entity']['notes']['woocommerce_order_id'];
+
+        $paymentId = $data['payload']['payment']['entity']['id'];
+
+        if (isset($data['payload']['payment']['subscription_id']) === true)
+        {
+            $this->processSubscription($orderId, $paymentId, false);
+        }
+
+        exit;
+    }
+
+    /**
+     * Handling the subscription charged webhook
+     *
+     * @param $data
+     */
+    protected function subscriptionCharged($data)
+    {
+        //
+        // Order entity should be sent as part of the webhook payload
+        //
+        $orderId = $data['payload']['subscription']['entity']['notes']['woocommerce_order_id'];
+
+        $this->processSubscription($orderId);
+
+        exit;
+    }
+
+    /**
+     * Helper method used to handle all subscription processing
+     *
+     * @param $orderId
+     * @param $paymentId
+     * @param $success
+     */
+    protected function processSubscription($orderId, $paymentId, $success = true)
+    {
+        //
+        // If success is false, automatically process subscription failure
+        //
+        if ($success === false)
+        {
+            return $this->processSubscriptionFailed($orderId);
+        }
+
+        // This method is used to process the subscription's recurring payment
+        $wcSubscription = wcs_get_subscriptions_for_order($orderId);
```

Comment:
@captn3m0 
We would need both 
```
$wcSubscriptionId = array_keys($wcSubscription)[0];

// We will only process one subscription per order
$wcSubscription = array_values($wcSubscription)[0];
```

both to passed here for the method to work. Also, regarding reusability, I agree, but we can say with surety that we process only one subscription at a time. So I think based on that constraint, this would be a valid assumption. 

Let us discuss this once in person and close it.